### PR TITLE
Rental model for storing PVF

### PR DESCRIPTION
**Pull request by Szegoo**
The PR aims to implement a rental payment mechanism for storing PVF for on-demand paras, reducing the initial cost of registering a para and paying for PVF storage

The plan is to have a smaller initial payment compared to regular PVF registration. However, an additional rent must be paid every `T::RentDuration` period.

In case the rent is not paid, the scheduler won't assign the para anymore, and it will be moved into the 'hibernation' stage. During this period, the PVF is still stored on-chain for an entire session in case it is needed for approval checking or disputes.

Once the para is no longer in the 'hibernation' stage, the 'prune_para' extrinsic can be called by anyone, and the caller will receive the initial deposit as a reward.

This PR is in its early draft stage and a work in progress.

This is part of: #1796 where more information can be found related to this.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4038474

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4039170

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4039442

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4042249

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4042384

**Comment by bkchr** [comment]
BTW, this should really be done as a RFC.

File: polkadot/runtime/common/src/paras_registrar/mod.rs [hunk 1]
```diff
@@ -431,6 +458,35 @@ pub mod pallet {
...
+		#[pallet::call_index(10)]
+		#[pallet::weight(<T as Config>::WeightInfo::register())]
+		pub fn pay_rent(origin: OriginFor<T>, id: ParaId) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let mut rent_info = RentedParas::<T>::get(id).ok_or(Error::<T>::NotParathread)?;
+			<T as Config>::Currency::reserve(&who, rent_info.rent_cost)?;
+
+			let now = shared::Pallet::<T>::session_index();
+			rent_info.last_rent_payment = now;
```
Comment (eskimor) [inline polkadot/runtime/common/src/paras_registrar/mod.rs:485]: What if the next rent gets paid, before the old expired? Shouldn't this be set to `now + remaining`?

File: polkadot/runtime/parachains/src/scheduler.rs [hunk 2]
```diff
@@ -395,15 +398,17 @@ impl<T: Config> Pallet<T> {
 					// the back of the claimqueue.
 					for drop in dropped_claims {
 						match T::AssignmentProvider::pop_assignment_for_core(core_idx, drop) {
-							Some(assignment) => {
+							Some(assignment)
```
Comment (eskimor) [inline polkadot/runtime/parachains/src/scheduler.rs:401]: Leaving this as a defensive check might make sense, but I feel like this should be checked earlier in the process. E.g. for on-demand an order should be deemed invalid if rent has not been paid.

File: polkadot/runtime/common/src/paras_registrar/mod.rs (hunk 1 above)
Comment (Szegoo) [reply polkadot/runtime/common/src/paras_registrar/mod.rs:485]: Agreed, this should be set to `now + remaining`. (NOTE: This is still an early draft :))

**Comment by Szegoo** [comment]
The associated RFC: https://github.com/polkadot-fellows/RFCs/pull/44

**Comment by bkchr** [comment]
Going to close based on the discussion from last week at the fellowship meetup.

**Pull request closed without merging**